drivers/syscon/syscon.c

```diff
@@ -12,106 +12,112 @@
 
 #include <drivers/syscon.h>
 
-static const struct device *syscon_dev;
+#include "syscon_common.h"
 
 struct syscon_generic_config {
 	DEVICE_MMIO_ROM;
-};
-
-static struct syscon_generic_config syscon_generic_config_0 = {
-	DEVICE_MMIO_ROM_INIT(DT_DRV_INST(0)),
+	uint8_t reg_width;
 };
 
 struct syscon_generic_data {
 	DEVICE_MMIO_RAM;
 	size_t size;
 };
 
-static struct syscon_generic_data syscon_generic_data_0;
-
-uintptr_t syscon_generic_get_base(void)
+static int syscon_generic_get_base(const struct device *dev, uintptr_t *addr)
 {
-	if (!syscon_dev) {
+	if (!dev) {
 		return -ENODEV;
 	}
 
-	return DEVICE_MMIO_GET(syscon_dev);
-}
-
-static int sanitize_reg(uint16_t *reg, size_t reg_size)
-{
-	/* Avoid unaligned readings */
-	*reg = ROUND_DOWN(*reg, sizeof(uint32_t));
-
-	/* Check for out-of-bounds readings */
-	if (*reg >= reg_size) {
-		return -EINVAL;
-	}
-
+	*addr = DEVICE_MMIO_GET(dev);
 	return 0;
 }
 
-int syscon_generic_read_reg(uint16_t reg, uint32_t *val)
+static int syscon_generic_read_reg(const struct device *dev, uint16_t reg, uint32_t *val)
 {
 	struct syscon_generic_data *data;
+	const struct syscon_generic_config *config;
 	uintptr_t base_address;
 
-	if (!syscon_dev) {
+	if (!dev) {
 		return -ENODEV;
 	}
 
-	data = syscon_dev->data;
+	data = dev->data;
+	config = dev->config;
 
 	if (!val) {
 		return -EINVAL;
 	}
 
-	if (sanitize_reg(&reg, data->size)) {
+	if (syscon_sanitize_reg(&reg, data->size, config->reg_width)) {
 		return -EINVAL;
 	}
 
-	base_address = DEVICE_MMIO_GET(syscon_dev);
+	base_address = DEVICE_MMIO_GET(dev);
 
 	*val = sys_read32(base_address + reg);
```

Comment:
I don't get it. Shouldn't this be `sys_read{8,16,32}` according to the value of `config->reg_width`?

Reply:
Done